### PR TITLE
Refactor Setup Command (take 2)

### LINKED ISSUE #594: Running `ghost setup` should be (more) idempotent
A user reported a case where they ended up with two systemd service files, one called `ghost_<ip address of blog>` and one called `ghost_<url of blog>`. Not for sure on how this happened, but based on the current code I have a theory:

- User runs `ghost setup`, gives ip address as url
- After a while, user runs `ghost setup` again, but gives the actual domain name this time

In this case, `ghost setup` _could_ potentially set up two different systemd service files, because the ghost instance "name" is generated by using the url (see [here](https://github.com/TryGhost/Ghost-CLI/blob/master/lib/commands/setup.js#L109)). Quick fix would probably be to skip the setup step that sets the instance name if the name is already set....however it's probably worth investigating further to see if there are more things in the setup command that can be fixed to make it idempotent.

If I find more things, I will add them here:

- [ ] Ensure instance name doesn't get reset if it already exists

### PR DESCRIPTION
**Pull request by acburdine**
closes #594, #588 

todo:
- [ ] manual testing with all the setup steps
- [x] fix unit tests

### CODE REVIEW

**Comment by acburdine** [comment]
@ErisDS @kirrg001 @vikaspotluri123 there are still some aforementioned todos, but would appreciate an initial code review as I _think_ I've gotten most of the actual refactoring done.

File: lib/commands/setup.js [hunk 1]
```diff
@@ -21,180 +21,144 @@ class SetupCommand extends Command {
...
 
-        if (argv.stages && argv.stages.length) {
-            const instance = this.system.getInstance();
+        const instance = this.system.getInstance();
+
+        // This is used so we can denote built-in setup steps
+        // and disable the "do you wish to set up x?" prompt
+        // We use symbols so extension-defined tasks can't
+        // mark themselves as internal and skip the prompt
+        const internal = Symbol('internal setup step');
```
Comment (acburdine) [inline lib/commands/setup.js]: as far as I can tell this is what symbols are _supposed_ to be used for, and it seemed like a good solution.

File: extensions/nginx/index.js [hunk 2]
```diff
@@ -25,23 +25,43 @@ class NginxExtension extends Extension {
...
+            name: 'SSL',
+            enabled,
+            task: (...args) => this.setupSSL(...args),
+            skip: ({tasks, ui}) => {
+                if (tasks.nginx.isSkipped()) {
+                    ui.log('Nginx setup task was skipped, skipping SSL setup', 'yellow');
+                    return true;
+                }
+
+                if (tasks.nginx.hasFailed()) {
```
Comment (acburdine) [inline extensions/nginx/index.js]: these two status checks could maybe be combined, seemed more prudent to leave the messaging different per case though.

File: extensions/mysql/index.js [hunk 3]
```diff
@@ -53,7 +51,8 @@ class MySQLExtension extends cli.Extension {
 
         return Promise.fromCallback(cb => this.connection.connect(cb)).catch((error) => {
             if (error.code === 'ECONNREFUSED') {
-                return Promise.reject(new cli.errors.ConfigError({
+                return Promise.reject(new ConfigError({
+
```
Comment (vikaspotluri123) [inline extensions/mysql/index.js]: Not really pertinent, but is this extra line needed?

Comment (acburdine) [reply extensions/mysql/index.js]: Not sure how that got added 🙃

**Comment by acburdine** [comment]
...aaaand it's done 🎉 

just needs manual testing & some commit squashing

**Comment by acburdine** [comment]
Note: I split off some of the commits in this PR into #840 -> it'll shrink this PR a little bit.

**Pull request merged**